### DISCUSSION TITLE
J-integral calculation for a 2D edge crack with collapse nodes at the crack tip

### DISCUSSION THREAD

**Discussion by tranm-ansto** [idaholab/moose#25044] [Q&A Modules: Solid mechanics] [answered]
Hello,

I am using the tensor mechanics module to calculate J-integral for a 2D edge crack with collapse nodes at the crack tip. 

I'm following the example in 
~\moose\modules\tensor_mechanics\test\tests\j_integral\j_integral_2d.i

The J-integral result at 50% through the width is shown below.
<img width="747" alt="center_crack_2d" src="https://github.com/idaholab/moose/assets/120076598/5bd39e51-b274-49c2-ae4d-b5f2e02a9db4">

I'm trying to follow this example to calculate J-integral for a 2D edge crack with collapse nodes at the crack tip. Figures of the 2D edge crack with web mesh are shown below.

<img width="252" alt="edge_crack_2d_1" src="https://github.com/idaholab/moose/assets/120076598/b2b07519-30fa-456a-b8c9-e9143e3055f5">

<img width="739" alt="edge_crack_2d_2" src="https://github.com/idaholab/moose/assets/120076598/dcff2302-4a85-48c3-b8f3-74d53f0a4295">

However, when I tried running the model, I got an error message about the "CrackFrontDefinition", shown below. 

<img width="1078" alt="edge_crack_2d_error" src="https://github.com/idaholab/moose/assets/120076598/711ca7fe-e40c-425f-afb0-d68aa714c8a9">

I tried selecting 1 of the collapse nodes at the crack tip. I also tried selecting all the collapse nodes. And, I still get the same error.

My input file is below.

```
[GlobalParams]
  order = FIRST
  family = LAGRANGE
  displacements = 'disp_x disp_y'
  volumetric_locking_correction = true
[]

[Mesh]
  file = edge_crack_2d.e
  # use_displaced_mesh = false
[]

[AuxVariables]
  [./SED]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[Modules/TensorMechanics/Master]
  [./master]
    strain = SMALL
    add_variables = true
    incremental = false
    generate_output = 'stress_xx stress_yy stress_zz vonmises_stress'
    planar_formulation = PLANE_STRAIN
  [../]
[]

[AuxKernels]
  [./SED]
    type = MaterialRealAux
    variable = SED
    property = strain_energy_density
    execute_on = timestep_end
  [../]
[]

[Functions]
  [./rampConstant]
    type = PiecewiseLinear
    x = '0. 1.'
    y = '0. 1.'
    scale_factor = -1
  [../]
[]

[BCs]
  [./Pressure]
    [./top_edge]
      boundary = 'topEdge'
      function = rampConstant
    [../]
  [../]

  [./crack_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'botEdge'
    value = 0.0
  [../]

  [./pin_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'pinNode'
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 207000
    poissons_ratio = 0.3
  [../]
  [./elastic_stress]
    type = ComputeLinearElasticStress
  [../]
[]

[DomainIntegral]
  integrals = JIntegral
  boundary = 'crackNode'
  crack_direction_method = CrackDirectionVector
  crack_direction_vector = '1 0 0'
  2d = true
  axis_2d = 2
  radius_inner = '0.501 0.505 0.510 0.515'
  radius_outer = '0.505 0.510 0.515 0.520'
  output_q = false
  incremental = false
[]



[Executioner]
  type = Transient

  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-ksp_gmres_restart'
  petsc_options_value = '101'

  line_search = 'none'

  l_max_its = 50
  nl_max_its = 20
  nl_rel_tol = 1e-12
  nl_abs_tol = 1e-5
  l_tol = 1e-2

  start_time = 0.0
  dt = 1

  end_time = 1
  num_steps = 1
[]

[Outputs]
  file_base = j_integral_edge_crack_2d
  exodus = true
  csv = true
[]
```

**Comment by GiudGiud** [comment]
Hello

can you plot only the ‘crackNode’ boundary in paraview to check that it meets the conditions?

Guillaume

**Reply by tranm-ansto** [reply]
Hi Guillaume,

Below is the 'crackNode' boundary plot.

![edge_crack_2d_3](https://github.com/idaholab/moose/assets/120076598/387dd458-98ec-40d2-b0ed-689eccaf1380)

I'm not quite sure what the conditions that need to be met are. Could you please provide a bit more detail on those conditions?

Minh

**Reply by GiudGiud** [reply]
The error message says the nodes should be aligned. How many nodes does « crackNode » have? And are they aligned? If it s all the nodes in the picture then they are not obviously

**Reply by GiudGiud** [reply]
Actually, if run you input with `--mesh-only` on the command line, the output to console should describe `crackNode` enough (the bounding box will be flat if all the nodes are aligned here)

**Comment by tranm-ansto** [comment, marked as answer]
It turns out that in the 'boundary' option accepts node set ID, not node set name. Please see below. The mesh seems fine. The code runs now. Thank you for the pointer, Guillaume.

```
[DomainIntegral]
  integrals = JIntegral
  boundary = 5000
  crack_direction_method = CrackDirectionVector
  crack_direction_vector = '1 0 0'
  2d = true
  axis_2d = 2
  radius_inner = '0.501 0.505 0.510 0.515 0.520 0.525'
  radius_outer = '0.505 0.510 0.515 0.520 0.525 0.530'
  output_q = false
  incremental = false
[]
```
